src/dpapimk_fmt_plug.c

```diff
@@ -0,0 +1,503 @@
+
+	p = strtokm(NULL, "*"); /* cipher algorithm */
+
+	p = strtokm(NULL, "*"); /* hash algorithm */
+
+	p = strtokm(NULL, "*"); /* pbkdf2 iterations */
+	cs.pbkdf2_iterations = (uint32_t) atoi(p);
+
+	p = strtokm(NULL, "*"); /* iv */
+	for (i = 0; i < 16; i++)
+		cs.iv[i] = atoi16[ARCH_INDEX(p[i * 2])] * 16
+			+ atoi16[ARCH_INDEX(p[i * 2 + 1])];
+
+	p = strtokm(NULL, "*"); /* encrypted length */
+	cs.encrypted_len = (uint32_t) atoi(p);
+
+	p = strtokm(NULL, "*"); /* encrypted stuff */
+	for (i = 0; i < cs.encrypted_len/2; i++)
+		cs.encrypted[i] = atoi16[ARCH_INDEX(p[i * 2])] * 16
+			+ atoi16[ARCH_INDEX(p[i * 2 + 1])];
+
+	MEM_FREE(keeptr);
+	return (void *)&cs;
+}
+
+
+static void set_salt(void *salt)
+{
+	cur_salt = (struct custom_salt *)salt;
+}
+
+static int decrypt_v1(unsigned char *key, unsigned char *iv, unsigned char *pwdhash, unsigned char *data)
+{
+	unsigned char out[MAX_CT_LEN+16];
+	unsigned char *last_key;
+	unsigned char *hmacSalt;
+	unsigned char *expected_hmac;
+	unsigned char computed_hmac[DIGEST_LEN1];
+	unsigned char encKey[DIGEST_LEN1];
+	DES_cblock ivec;
+	DES_key_schedule ks1, ks2, ks3;
+
+	memset(out, 0, sizeof(out));
+	DES_set_key((DES_cblock *) key, &ks1);
+	DES_set_key((DES_cblock *) (key + 8), &ks2);
+	DES_set_key((DES_cblock *) (key + 16), &ks3);
+	memcpy(ivec, iv, 8);
+	DES_ede3_cbc_encrypt(data, out, (cur_salt->encrypted_len / 2) * 8, &ks1, &ks2, &ks3, &ivec,  DES_DECRYPT);
+
+	hmacSalt = out;
+	expected_hmac = out + 16;
+	last_key = out + (cur_salt->encrypted_len / 2) - 64;
+
+	hmac_sha1(pwdhash, 32, hmacSalt, 16, encKey, DIGEST_LEN1);
+	hmac_sha1(encKey, DIGEST_LEN1, last_key, 64, computed_hmac, DIGEST_LEN1);
+
+	return memcmp(expected_hmac, computed_hmac, DIGEST_LEN1);
+}
+
+static int decrypt_v2(unsigned char *key, unsigned char *iv, unsigned char *pwdhash, unsigned char *data)
+{
+	unsigned char out[MAX_CT_LEN+16];
+	unsigned char *last_key;
+	unsigned char *hmacSalt;
+	unsigned char *expected_hmac;
+	unsigned char hmacComputed[DIGEST_LEN2];
+	unsigned char encKey[DIGEST_LEN2];
+
+	AES_KEY aeskey;
+
+	AES_set_decrypt_key(key, KEY_LEN2 * 8, &aeskey);
+	AES_cbc_encrypt(data, out, (cur_salt->encrypted_len / 2) * 8, &aeskey, iv, AES_DECRYPT);
+
+	hmacSalt = out;
+	expected_hmac = out + 16;
+	last_key = out + (cur_salt->encrypted_len / 2) - 64;
+
+	hmac_sha512(pwdhash, 20, hmacSalt, IV_LEN2, encKey, DIGEST_LEN2);
+	hmac_sha512(encKey, DIGEST_LEN2, last_key, 64, hmacComputed, DIGEST_LEN2);
+
+	return memcmp(expected_hmac, hmacComputed, DIGEST_LEN2);
+}
+
+
+static int crypt_all(int *pcount, struct db_salt *salt)
+{
+	const int count = *pcount;
+	int index = 0;
+
+#ifdef _OPENMP
+#pragma omp parallel for
+#endif
+	for (index = 0; index < count; index += MAX_KEYS_PER_CRYPT) {
+		unsigned char *passwordBuf;
+		int passwordBufSize, digestlen = 20;
+		unsigned char out[MAX_KEYS_PER_CRYPT][KEY_LEN2 + IV_LEN2];
+		unsigned char out2[MAX_KEYS_PER_CRYPT][KEY_LEN2 + IV_LEN2];
+
+		SHA_CTX ctx;
+		MD4_CTX ctx2;
+
+		int i;
```

Comment:
Tighten line spacing. Various variable declarations can be clubbed together without newlines.